MLJ Integration

tylerjthomas9 opened the pull request:
In reference to https://github.com/beacon-biosignals/CatBoost.jl/issues/9

I added an MLJ Interface for `CatBoostRegressor`. I am not sure what the best way to deal with the python vs MLJ interfaces (we could just put them in separate modules inside the package). This was just a quick port of `CatBoostRegressor` MLJ, so I am sure it'll change with time. Let me know what you think

ablaom commented:
FYI: https://github.com/JuliaAI/MLJTestInterface.jl

ablaom commented:
Like this: https://github.com/JuliaAI/MLJXGBoostInterface.jl/blob/a204646a3f367652b22ffd9497e4a0bc1b773b6e/test/runtests.jl#L195

ablaom commented:
Happy to provide a code review for the MLJ interface if required (although may have to wait until Jan). Ping me when you're ready.

tylerjthomas9 commented:
@ablaom I would love if you looked at the MLJ interface when you get a chance.

The summary of the changes that I have made is that the core interface is the same, but there is a submodule, `MLJCatBoostInterface` that has all the MLJ code. 

Here are some items that I have questions about:
- `using CatBoost` and `using CatBoost.MLJCatBoostInterface` both export `CatBoostRegressor` and `CatBoostClassifier`. I do not picture a user having issues with this, but it doesn't seem right. My first thought was to rename the base interface versions to `PyCatBoostRegressor` and ``PyCatBoostClassifier`, because they are the actual python classes, and not structs.  However, I am not sure this breaking change is necessary. 
- I am getting MLJ data type warnings on CatBoostClassifier. Do you know why these are appearing?

ablaom commented:
@tylerjthomas9 Thanks for this work. I shall take a look at it shortly. 

> using CatBoost and using CatBoost.MLJCatBoostInterface both export CatBoostRegressor and 

There is no need `export` the MLJ models, so if I understand right, they don't need different names. For example, when you use `@load` in MLJ you only import the module-qualified version of CatBoostRegressor, which will not conflict with the CatBoost if, for some reason that has been imported. 

> I am getting MLJ data type warnings on CatBoostClassifier. Do you know why these are appearing?v

I'll need more context but will possibly get this when I start to review.

tylerjthomas9 commented:
> I'll need more context but will possibly get this when I start to review.

I get the following warning when running the MLJ test interface:

```
┌ Warning: The number and/or types of data arguments do not match what the specified model
│ supports. Suppress this type check by specifying `scitype_check_level=0`.
│ 
│ Run `@doc CatBoost.jl.CatBoostClassifier` to learn more about your model's requirements.
│ 
│ Commonly, but non exclusively, supervised models are constructed using the syntax
│ `machine(model, X, y)` or `machine(model, X, y, w)` while most other models are
│ constructed with `machine(model, X)`.  Here `X` are features, `y` a target, and `w`
│ sample or class weights.
│ 
│ In general, data in `machine(model, data...)` is expected to satisfy
│ 
│     scitype(data) <: MLJ.fit_data_scitype(model)
│ 
│ In the present case:
│ 
│ scitype(data) = Tuple{Table{AbstractVector{Count}}, AbstractVector{Count}}
│ 
│ fit_data_scitype(model) = Tuple{Union{Table{<:Union{AbstractVector{<:Continuous}, AbstractVector{<:Count}, AbstractVector{<:OrderedFactor}}}, AbstractMatrix{Continuous}}, Union{AbstractVector{<:Finite}, AbstractVector{<:Continuous}}} 
└ @ MLJBase C:\Users\TylerThomas\.julia\packages\MLJBase\9Nkjh\src\machines.jl:230
```

I have tried playing with the input scitypes quite a bit, but I can't get it to resolve. I may be missing something obvious here.

ablaom commented on examples/mlj/binary.jl:
```diff
@@ -0,0 +1,29 @@
+# Initialize data
+cat_features = [0, 1]
+train_data = DataFrame([["a", "a", "c"], ["b", "b", "d"], [1, 4, 30], [4, 5, 40],
+                        [5, 6, 50], [6, 7, 60]], :auto)
```
This is generating a warning because `scitype(["a", "a", "c"])` is `Textual` but, according to your implementation's `input_scitype`, only `Continuous`, `OrderedFactor`, or `Count` columns are allowed. 

Now you have specified `cat_features = [0, 1]` (more on this hyper-parameter in my comments below) from which I infer you mean `["a", "a", "c"]` to be categorical, presumably `Multiclass` and not `OrderedFactor`. So you probably want to replace with `coerce(["a", "a", "c"], Multiclass)` in this test (`coerce` is exported by `ScientificTypes` and also `MLJBase`).

tylerjthomas9 replied:
Thank you for explaining this. I understand now. This should be resolved in my latest commit.

ablaom commented on test/mlj_interface.jl:
```diff
@@ -0,0 +1,60 @@
+
+    @testset "CatBoostClassifier" begin
+        X = DataFrame(; a=[1, 4, 5, 6], b=[4, 5, 6, 7])
+        y = [0, 0, 1, 1]
```
You are generating a warning here because `scitype(y) = AbstractVector{Count}` but your implementation declares `target_scitype(y)` = `AbstractVector{<:Finite}` (which makes sense because your model is a classifier). 

So you probably want to do `y = coerce([0, 0, 1, 1], OrderedFactor)` or `y = coerce([0, 0, 1, 1], Multiclass)` here.

tylerjthomas9 replied:
Thank you for explaining this.

ablaom reviewed (commented):
Looks like a lot of work has put into this, thanks! So far, I am only addressing the scitype issues in this review and would like a further opportunity to review more carefully later. 

**1** My first objection is that you expose the hyper-parameters `cat_features` and `text_feautres` in the MLJ models. In MLJ we avoid hyper-parameters that pass characteristics of the data. The data should speak for itself. And this is not a problem because MLJ requires the user to represent data in a way that allows inference of what features are categorical from their `scitype`. In other words, `cat_features` and `text_feautres` (and maybe some others??) are things your interface should be determining from the data presented internally. 

I'm guessing that any input column that has `OrderedFactor` scitype you will want to internally convert to integer, for  catboost (which you can do with `MLJModelInterface.int`) and that any input column that has `Multiclass` scitype you will flag internally using `cat_features` - and maybe you can leave this, but probably you need convert to integers for catboost also. Any `Count` features are integers and are not to be treated as categorical but as  "numerical" which catboost will do automatically. 

I'm not sure what the `text_features` hyperparameter does in catboost. Does catboost support some kind of semantic anaysis? Or is this just a different kind of categorical feature? You'll have to explain to me how catboost interprets such features for me to comment further. But, as far as the `String` types I'm seeing in the test examples,  I think ordinary `Multiclass` is intended, and so those string types need coercion to categorical vectors in the test scripts, as I have indicated.

**2.** It seems strange to me that your `target_scitype` declaration for the classifier includes `Continuous`. Is there a reason for this?

tylerjthomas9 commented:
Thank you for taking the time to review. 

> **2.** It seems strange to me that your `target_scitype` declaration for the classifier includes `Continuous`. Is there a reason for this?

Oops. I accidentally added this. I have removed this. 

> **1** My first objection is that you expose the hyper-parameters `cat_features` and `text_feautres` in the MLJ models. In MLJ we avoid hyper-parameters that pass characteristics of the data. The data should speak for itself. And this is not a problem because MLJ requires the user to represent data in a way that allows inference of what features are categorical from their `scitype`. In other words, `cat_features` and `text_feautres` (and maybe some others??) are things your interface should be determined from the data presented internally.

I have added automatic detection of these columns using the function `prepare_input`. Let me know what you think.

> any input column that has `Multiclass` scitype you will flag internally using `cat_features` - and maybe you can leave this, but probably you need convert to integers for catboost also. Any

I think that we can leave this column. Catboost should be able to handle it. 

> I'm not sure what the `text_features` hyperparameter does in catboost. Does catboost support some kind of semantic anaysis?

I am not too familiar with how catboost handles text features, but it seems like the text is split on spaces, then converted to numeric features using Bag of Words. Here is a little more info if you're interested: https://catboost.ai/en/docs/concepts/algorithm-main-stages_text-to-numeric.

ablaom commented:
> I am not too familiar with how catboost handles text features, but it seems like the text is split on spaces, then converted to numeric features using Bag of Words. 

Ah ha! In that case you may want to consider enlarging `input_scitype` for both models to include columns with elscitype `Textual` and pass such columns onto catboost as is, but with the relevant column flagged in `text_features`, which you internally set, like `cat_features`.  It will be important to document carefully in the doc string that `String` data will be interpreted as `Textual` and not `Multiclass`. Some users are likely going to use strings, thinking they are going to be interpreted as `Multiclass`.

ablaom commented on examples/mlj/binary.jl:
```diff
@@ -6,24 +6,25 @@ using MLJBase
 
 # Get predicted classes
-preds_class = MLJBase.predict(mach, eval_data)
+preds_class = MLJBase.predict_mean(mach, eval_data)
```
Do you mean `predict_mode` here? This is a classifier.

tylerjthomas9 replied:
Yes it should. I have made this change.

ablaom commented on examples/mlj/multiclass.jl:
```diff
@@ -0,0 +1,31 @@
+# Initialize data
+train_data = DataFrame([coerce(["a", "a", "c"], MLJBase.Multiclass),
+                        coerce(["b", "b", "d"], MLJBase.Multiclass),
+                        coerce([0, 0, 1], OrderedFactor), [4, 5, 40], [5, 6, 50],
```
Maybe for consistency make this `MLJBase.OrderedFactor`

tylerjthomas9 replied:
Done!

ablaom commented on src/MLJCatBoostInterface.jl:
```diff
@@ -28,6 +28,40 @@ function mlj_to_kwargs(model)
+- `Vector{Int64}`
+"""
+function get_dtype_feature_ix(X, dtype)
+    return findall(eltype.(MMI.scitype.(eachcol(X))) .<: dtype)
```
Perhaps simpler is `return findall(MMI.schema(X).scitypes .< dtype)`

tylerjthomas9 replied:
Yes, that is simpler. Thank you! I have added it in the latest commit.

ablaom commented on src/MLJCatBoostInterface.jl:
```diff
@@ -28,6 +28,40 @@ function mlj_to_kwargs(model)
+"""
+function prepare_input(X)
+    order_factor_ix = get_dtype_feature_ix(X, OrderedFactor)
+    for col in names(X)[order_factor_ix]
```
What is this method `names` ? I think the method of that name in Tables.jl is only for named tuple tables, and matrix tables. You could do `MMI.schema(X).names`, which amounts to something like `Tables.schema(Tables.rowtable(X)).names` or `Tables.columnnames(Tables.rowtable(X))`. Unfortunately Tables.jl does not have a `names` method that can be directly applied to any table, as I understand it.

ablaom commented on src/MLJCatBoostInterface.jl:
```diff
@@ -28,6 +28,40 @@ function mlj_to_kwargs(model)
+    end
+
+    cat_features = get_dtype_feature_ix(X, Multiclass) .- 1 # convert to 0 based indexing
+    text_features = get_dtype_feature_ix(X, AbstractString) .- 1 # convert to 0 based indexing
```
I think you want a scitype here: replace `AbstractString -> Textual`.

ablaom replied:
Okay, you already did this. Resolved.

ablaom commented on src/MLJCatBoostInterface.jl:
```diff
@@ -28,6 +28,40 @@ function mlj_to_kwargs(model)
+    for col in names(X)[order_factor_ix]
+        X[:, col] = MMI.int(X[:, col])
+    end
+
```
As I understand the later call to this function, we should only be assuming that `X` satisfies `Tables.istable(X) =true`, ie, we don't know what kind of table this is. So I'd think `names(X)` should probably be something like `MMI.schema(X).names` or `Tables.columnnames(columns(X))` or something like that. And you will not be able to mutate columns using the line at 56, but (unfortunately) must create a new object (eg a named-tuple of vectors, which is also a Tables.jl table) by grabbing each column using the [Tables API](https://tables.juliadata.org/stable/#Using-the-Interface-(i.e.-consuming-Tables.jl-compatible-sources)).

ablaom replied:
resolved

ablaom commented on test/mlj_interface.jl:
```diff
@@ -42,15 +42,19 @@
-            failures, summary = MLJTestInterface.test([CatBoostRegressor],
-                                                      MLJTestInterface.make_regression()...;
+            data = MLJTestInterface.make_regression()
+            X = DataFrame(data[1])
```
You shouldn't need to convert to a `DataFrame` if you make the genericity changes explained above. Your interface must work for any table (with `Tables.istable(X) = true`).

ablaom reviewed (commented):
Nice progress. 

Heads up regarding all this data preprocessing you are implementing: You probably should leave this for a future PR, but ultimately we are going to want to:

- Implement `MMI.update`
- Implement an MLJ data ["front-end"](https://alan-turing-institute.github.io/MLJ.jl/dev/adding_models_for_general_use/#Implementing-a-data-front-end) where all the preprocessing of data (converting MLJ user provided table into pandas-ready object) is factored out of `fit`/`update`. 

Then iteration of the models can be  efficiently and externally controlled using `MLJIteration.jl`, which would be pretty cool. Fortunately, we have EvoTrees.jl as a guide.

tylerjthomas9 commented on examples/mlj/multiclass.jl:
```diff
@@ -0,0 +1,31 @@
+MLJBase.fit!(mach)
+
+# Get predicted classes
+preds_class = MLJBase.predict_mean(mach, eval_data)
```
Done!

ablaom commented on src/MLJCatBoostInterface.jl:
```diff
@@ -51,16 +51,18 @@ Get cat features for model
+    order_factor_ix = get_dtype_feature_ix(table_input, OrderedFactor)
+    new_columns = Dict([col => Vector{Int64}(table_input[col]) for col in
+                                                                   columns[order_factor_ix]]...)
+    table_input = (; table_input..., new_columns...)
```
Don't you want to drop the columns for which you have just computed replacements? Or am I missing something?

tylerjthomas9 commented:
I took a stab at adding `MMI.update` and the MLJ data "front-end". Let me know what you think. I wasn't sure what to do with `cat_features`, `text_features`, but this seems to work.

ablaom commented on src/MLJCatBoostInterface.jl:
```diff
@@ -54,11 +59,10 @@ function prepare_input(X)
-    new_columns = Dict([col => MMI.int(table_input[col]) for col in
-                                                             columns[order_factor_ix]]...)
+    order_factor_cols = columns[get_dtype_feature_ix(table_input, OrderedFactor)]
+    new_columns = Dict([col => MMI.int(table_input[col]) for col in order_factor_cols]...)
```
I'm a bit nervous about use of dictionary here, because the order is unpredictable. This order conceivably matters: In trees, when determining which feature to split on there can be ties (draws). Maybe in that case the order of the features matters??

Is there a good reason for not using a named tuple here?

ericphanson replied:
We already depend on OrderedCollections for OrderedDict's btw, so we could use that

tylerjthomas9 replied:
There was not a good reason. I have updated it to a named tuple.

ablaom commented on src/MLJCatBoostInterface.jl:
```diff
@@ -72,12 +72,30 @@ end
-const CATBOOST_MODELS = Union{CatBoostClassifier,CatBoostRegressor}
+const CatBoostModels = Union{CatBoostClassifier,CatBoostRegressor}
+
+function MMI.selectrows(::CatBoostModels, I, X, y)
```
The "data " part of this signature should match the new signature of `fit`. So you are missing the two extra items `cat_features` and `text_features`, which you will return untouched as the last two return values.

tylerjthomas9 replied:
Done!

tylerjthomas9 commented on src/MLJCatBoostInterface.jl:
```diff
@@ -74,9 +75,13 @@ include("mlj_catboostregressor.jl")
+    return X.iloc[py_I,], y[py_I], cat_features, text_features
+end
+
+function MMI.selectrows(::CatBoostModels, I::Colon, X, y, cat_features, text_features)
```
I added another version of select rows for when we pass a Colon. Pandas can't handle that as the indexing. I thought that just dispatching 2 versions was a good way to handle this.

ablaom replied:
Good catch!

ablaom commented on src/MLJCatBoostInterface.jl:
```diff
@@ -72,12 +72,30 @@ end
+    if current_iterations < mlj_model.iterations
+        iterations = mlj_model.iterations - current_iterations
+        fitresult.fit(X, y; init_model=fitresult, iterations=iterations)
+    else
```
You should check that `iterations` is the *only* hyperparameter that has changed. To that end I suggest that `fit` (and `update`) makes a deepcopy of `model`, which it returns as `cache` (2nd return value), which MLJ will ensure gets passed as `cache` in the `update` call being implemented here.  You can then compare the current `model` with the last one used in any training, `cache` (and also have access to previous `iterations` without counting trees in the `fitresult`). MLJModelInterface has a convenience function, [`is_same_except`](https://github.com/JuliaAI/MLJModelInterface.jl/blob/d9e9703947fc04b0a5e63680289e41d0ba0d65bd/src/equality.jl#L74)  to make the comparison easy. You can see it applied [here](https://github.com/FluxML/MLJFlux.jl/blob/452c09d7dc13914f4057c661448bc310c9362d3d/src/mlj_model_interface.jl#L115).

tylerjthomas9 replied:
I am a little stumped on the best way to do this. `is_same_except` doesn't work with the python objects. I have updated `MMI.update` to create a brand new CatBoost model to start training where the last one left off. (It doesn't like resuming an already trained model, so I have to create a new one and initialize it with the previous one).

ablaom replied:
> I am a little stumped on the best way to compare hyperparameters. is_same_except doesn't work with the python objects. There are a few differences between the two models, Should I compare the two parameter dictionaries?

I don't understand what python has to do with it. You are comparing two Julia structs, with ordinary fields no?

Okay, i just had a look at your [Julia structs](https://github.com/beacon-biosignals/CatBoost.jl/blob/45647592ed06c43d7267f5622c7fdc2399796ecc/src/mlj_catboostclassifier.jl#L2) for the model hyperparameters and see you are doing something very weird. See my other [comment](https://github.com/beacon-biosignals/CatBoost.jl/pull/16#discussion_r1082047909). MLJ cannot control an iterative model where the default value of the iteration parameter is `nothing`. It needs to have a value. And it cannot change its value except by passing an instance with the new value.

tylerjthomas9 replied:
Ah, I see. I updated it [here](https://github.com/beacon-biosignals/CatBoost.jl/pull/16/commits/3f961b7772e08ee204a7074230f0a3a3adadf2d9) to compare the mlj model and the cache version from the previous update.

ablaom replied:
resolved

ablaom reviewed (commented):
Regarding the new "data front-end". You should also factor out the preprocessing you have in `predict` into a `reformat` with signature `reformat(::CatBoostClassifier, Xnew)` which returns `(X_preprocessed,)` (return value must be a tuple!) and a corresponding `selectrows(::CatBoostClassifier, I, Xprocessed)` whose output again get's wrapped as a 1-tuple.

And then, presumably you want to do the same for the regressor (perhaps it suffices to extend the types of the methods already written here for classification).

ablaom commented on src/mlj_catboostclassifier.jl:
```diff
@@ -0,0 +1,134 @@
+
```
What are all these `nothing`s ? Why aren't proper default values being set here?

tylerjthomas9 replied:
I wanted it to default to the python library's default parameters. My thought process was that it was lower maintenance, and it avoids getting different default parameters if the python version changes things up. 

However, I'll gladly add them in if you think they should be included.

ablaom replied:
Yeah, I realize this is a maintenance headache. I've had the same complaint about XGBoost. But this seems particular to the wrapped gradient tree boosters - other models don't have many parameters and defaults rarely change. MLJ kind of needs this, at least at the moment, for multiple reasons. For example, it's good to know what the default type of a field value is when constructing hyperparameter ranges for tuning. 

So, can you add them please? 🙏🏾

tylerjthomas9 replied:
I have added them. There are quite a few dynamic parameters in CatBoost that don't have universal default values. I have kept them as `nothing` for now.

ablaom replied:
Yeah that should be okay.

ablaom commented on src/mlj_docstrings.jl:
```diff
@@ -0,0 +1,146 @@
+const CATBOOST_PARAMS_DOCS_LINK = "https://catboost.ai/en/docs/references/training-parameters/"
+
+"""
+$(MMI.doc_header(CatBoostRegressor))
```
```suggestion
$(MMI.doc_header(CatBoostClassifier))
```

ablaom replied:
Resolved.

ablaom commented on src/mlj_docstrings.jl:
```diff
@@ -0,0 +1,146 @@
+
+
+# Hyper-parameters
+#TODO: Document hyper-parameters
```
I think given the number of parameters and the associated maintenance burden, it suffices to refer to the catboost.ai docs for the details. If there are hyperparameter values which would differ (e.g., symbol expected instead of string, or the value is an instance of some CatBoost.jl custom type) then these should be noted. But maybe this has been avoided?

ericphanson replied:
One option is to interpolate in the python docs if there's some appropriate ones. That's what we did here: https://github.com/beacon-biosignals/CatBoost.jl/blob/9cab2d6ebce3cf205923a26140b4bda2443907b3/src/CatBoost.jl#L44-L46

Not perfect but I think it's helpful still

ablaom commented on src/mlj_docstrings.jl:
```diff
@@ -0,0 +1,146 @@
+
+# Operations
+
+- `predict(mach, Xnew)`: return predictions of the target given new
```
It seems the two method descriptions are switched:

```suggestion
- `predict(mach, Xnew)`: return probabilistic predictions of the target given new
```

I don't see that you need to document `predict_mode` (we haven't done so elsewhere) but won't insist you don't.

tylerjthomas9 replied:
I removed it

ablaom commented on src/mlj_docstrings.jl:
```diff
@@ -0,0 +1,146 @@
+
+# Report
+
+- `feature_importances`: DataFrame of feature importances
```
As you are exposing feature importances, you may also want to implement `MMI.feature_importances(model::M, fitresult, report)` and the associated trait `MLJModelInterface.reports_feature_importances(::Type{<:M})`. This will  allow planned feature selection methods (eg. recursive feature elimination) to use your models.

tylerjthomas9 replied:
Done!

ablaom commented on src/mlj_docstrings.jl:
````diff
@@ -0,0 +1,146 @@
+```
+using CatBoost
+using DataFrames
+using MLJBase
````
```suggestion
using MLJ
```

ablaom commented on src/mlj_docstrings.jl:
````diff
@@ -0,0 +1,146 @@
+
+```
+using CatBoost
+using DataFrames
````
```suggestion
```

ablaom commented on src/mlj_docstrings.jl:
```diff
@@ -0,0 +1,146 @@
+  `Textual` columns will be passed to catboost as `text_features`,
+  `Multiclass` columns will be passed to catboost as `cat_features`, and
+  `OrderedFactor` columns will be converted to integers.
+
```
nice

ablaom commented on src/mlj_docstrings.jl:
```diff
@@ -0,0 +1,146 @@
+using DataFrames
+using MLJBase
+
+X = DataFrame(; a=[1, 4, 5, 6], b=[4, 5, 6, 7])
+y = [0, 0, 1, 1]
```
```suggestion
X = (
    duration = [1.5, 4.1, 5.0, 6.7], 
    n_phone_calls = [4, 5, 6, 7], 
    department = coerce(["acc", "ops", "acc", "ops"], Multiclass), 
)
y = coerce([0, 0, 1, 1], Multiclass)
```

ablaom commented on src/mlj_docstrings.jl:
```diff
@@ -0,0 +1,146 @@
+y = [0, 0, 1, 1]
+
+model = CatBoost.CatBoostClassifier(iterations=5)
+mach = machine(model, X, y)
```
This fails. Here `model` does not appear to be a valid MLJ model but has type `PythonCall.Py`.

tylerjthomas9 replied:
Ah, this is from before I put the MLJ models in their own module. I will update these examples.

ablaom commented on src/mlj_docstrings.jl:
````diff
@@ -0,0 +1,146 @@
+mach = machine(model, X, y)
+MLJBase.fit!(mach)
+preds = MLJBase.predict(mach, X)
+```
````
You may want to rework this example along the lines suggested for the regressor.

ablaom reviewed (commented):
Phew! I think we're getting close to the finish line here. 

Just some doc string fixes and suggestions here. 

I'd still like to run some `MLJTestIntegration.jl` tests, just to be sure. This might have to wait till next week. I've gotten busy elsewhere. 

Probably this is a PythonCall issue, but I sometimes get complaints about missing python modules when testing your code. This usually requires me to start over with a new environment, which is a pain, or some other random stuff I can't put my finger on. Do have any advice about this?

tylerjthomas9 commented:
> This might have to wait till next week. I've gotten busy elsewhere.

Absolutely no rush. Thank you again for taking the time to review and help with this PR

tylerjthomas9 commented:
> Probably this is a PythonCall issue, but I sometimes get complaints about missing python modules when testing your code. This usually requires me to start over with a new environment, which is a pain, or some other random stuff I can't put my finger on. Do have any advice about this?

> Probably this is a PythonCall issue, but I sometimes get complaints about missing python modules when testing your code. This usually requires me to start over with a new environment, which is a pain, or some other random stuff I can't put my finger on. Do have any advice about this?

That is odd, I have not gotten issues like this, but someone reported having import issues here: https://github.com/beacon-biosignals/CatBoost.jl/issues/17#issuecomment-1387015936. I think the best thing to do would be to manually add catboost via CondaPkg (`CondaPkg.add("catboost", channel="conda-forge")`). I do not know what causes this to randomly occur, but it might be a PythonCall issue.

ablaom commented:
Mmmm..

```
signal (11): Segmentation fault: 11
in expression starting at REPL[27]:1
Allocations: 247400422 (Pool: 247300932; Big: 99490); GC: 123

Process julia-cat segmentation fault: 11
```

Are you ever seeing this with CatBoost usage?

tylerjthomas9 commented:
> Are you ever seeing this with CatBoost usage?

I have not ever gotten this. Are you on Mac? I found this potentially relevant issue: https://github.com/catboost/catboost/issues/2201

Do you know which test is causing this?

ablaom commented:
I'm getting the fail reproducibly when I run MLJTestIntegration. Investigating further,  I have found some regular bugs with the MLJ interface. Let's start with those and come back to the seg fault. 

Here's one to start:

```julia
using CatBoost.MLJCatBoostInterface
using Tables

X = Tables.table(rand(10,2))
model = CatBoostClassifier()
selectrows(model, X, [1, 2, 3])

# julia> selectrows(model, X, [1, 2, 3])
# ERROR: type Array has no field slice
# Stacktrace:
#  [1] getproperty
#    @ ./Base.jl:38 [inlined]
#  [2] selectrows(#unused#::CatBoostClassifier, I::Tables.MatrixTable{Matrix{Float64}}, data_pool::Vector{Int64})                                                                   
#    @ CatBoost.MLJCatBoostInterface ~/Julia/CatBoost/src/MLJCatBoostInterface.jl:105
#  [3] top-level scope
#    @ REPL[27]:1
```

ablaom commented on src/MLJCatBoostInterface.jl:
```diff
@@ -0,0 +1,159 @@
+
+function MMI.selectrows(::CatBoostModels, I, data_pool)
+    py_I = numpy.array(numpy.array(I))
+    return (data_pool.slice(py_I))
```
Something wrong here. Why should `data_pool` have `slice` as field? See my other comment. Also, I think your missing a comma:
```suggestion
    return (data_pool.slice(py_I),)
```

tylerjthomas9 replied:
Should `selectrows` handle data in the format that `MMI.update`/`MMI.fit` uses? I can add methods for handling generic table subsetting too. I have slice here because I assume that the data is in a catboost pool at this point, and you have to slice the pool to subset it. 

https://catboost.ai/en/docs/concepts/python-reference_pool_slice

ablaom replied:
Yes you are right! That's not the issue.  But I have this error, and I think the comma is the problem.

```julia
using MLJ
model = CatBoostClassifier()
mach = machine(model, X, y)
train, test = partition(eachindex(y), 0.7)

julia> fit!(mach, rows=train)
[ Info: Training machine(CatBoostClassifier(iterations = 1000, …), …).
┌ Error: Problem fitting the machine machine(CatBoostClassifier(iterations = 1000, …), …). 
└ @ MLJBase ~/.julia/packages/MLJBase/9kTjj/src/machines.jl:682
[ Info: Running type checks... 
[ Info: Type checks okay. 
ERROR: Python: TypeError: 'Pool' object is not iterable
Python stacktrace: none
Stacktrace:
 [1] pythrow()
   @ PythonCall ~/.julia/packages/PythonCall/3GRYN/src/err.jl:94
 [2] errcheck
   @ ~/.julia/packages/PythonCall/3GRYN/src/err.jl:10 [inlined]
 [3] pyiter(x::PythonCall.Py)
   @ PythonCall ~/.julia/packages/PythonCall/3GRYN/src/abstract/iter.jl:6
 [4] iterate(x::PythonCall.Py)
   @ PythonCall ~/.julia/packages/PythonCall/3GRYN/src/Py.jl:339
 [5] fit_only!(mach::Machine{CatBoostClassifier, true}; rows::Vector{Int64}, verbosity::Int64, force::Bool, composite::Nothing)                                                     
   @ MLJBase ~/.julia/packages/MLJBase/9kTjj/src/machines.jl:680
 [6] #fit!#63
   @ ~/.julia/packages/MLJBase/9kTjj/src/machines.jl:778 [inlined]
 [7] top-level scope
   @ REPL[29]:1
```

tylerjthomas9 replied:
added the missing comma

ablaom replied:
resolved

ablaom commented:
There's still something wrong with that `selectrows` implementation. Here's a MWE:

```
using CatBoost.MLJCatBoostInterface
using MLJBase

model = CatBoostClassifier()
X, y = make_blobs()
train, test = partition(eachindex(y), 0.7)

Xpy = MLJBase.reformat(model, X)

# works fine:
selectrows(model, 71:99, Xpy...)

# fails:
selectrows(model, 71:100, Xpy...)
# ERROR: Python: CatBoostError: catboost/libs/helpers/array_subset.h:605: TIndexedSubset[29] (100) >= srcSize (100)
# Python stacktrace:
#  [1] _catboost._PoolBase._take_slice
#    @ _catboost.pyx:4425
#  [2] _catboost._PoolBase._take_slice
#    @ _catboost.pyx:4419
#  [3] slice
#    @ catboost.core ~/.julia/environments/catboost/.CondaPkg/env/lib/python3.10/site-packages/catboost/core.py:1052
# Stacktrace:
#  [1] pythrow()
#    @ PythonCall ~/.julia/packages/PythonCall/3GRYN/src/err.jl:94
#  [2] errcheck
#    @ ~/.julia/packages/PythonCall/3GRYN/src/err.jl:10 [inlined]
#  [3] pycallargs(f::PythonCall.Py, args::PythonCall.Py)
#    @ PythonCall ~/.julia/packages/PythonCall/3GRYN/src/abstract/object.jl:210
#  [4] pycall(f::PythonCall.Py, args::PythonCall.Py; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})                                        
#    @ PythonCall ~/.julia/packages/PythonCall/3GRYN/src/abstract/object.jl:228
#  [5] pycall
#    @ ~/.julia/packages/PythonCall/3GRYN/src/abstract/object.jl:218 [inlined]
#  [6] #_#11
#    @ ~/.julia/packages/PythonCall/3GRYN/src/Py.jl:352 [inlined]
#  [7] Py
#    @ ~/.julia/packages/PythonCall/3GRYN/src/Py.jl:352 [inlined]
#  [8] selectrows(#unused#::CatBoostClassifier, I::UnitRange{Int64}, data_pool::PythonCall.Py)
#    @ CatBoost.MLJCatBoostInterface ~/Julia/CatBoost/src/MLJCatBoostInterface.jl:105
#  [9] top-level scope
#    @ REPL[41]:1
```
0-based indexing adjustment?

tylerjthomas9 commented:
Ah, you're right. I just pushed a fix.

ablaom commented:
Thanks for that progress.

My stacking integration tests are failing and I have a pretty good idea what the problem is. In these tests, the folds can get quite small, and you have folds where the target (iris in this case) is missing a class. It's the responsibility of the implementation to track all classes, even those not explicitly manifest.

Currently this tracking in not happening, because we have `pool=missing` in [this line](https://github.com/tylerjthomas9/CatBoost.jl/blob/40ca6c9e563331f7c7a8c4aa5b2d2a1e6a276840/src/mlj_catboostclassifier.jl#L93). This means only the actual classes that python knows about are tracked. We only need to specify the correct pool and the `UnivariateConstructor` will automatically add zero probabilities for the classes not seen (and, so in stacking, always return a probability array of the correct dimension to the meta learner, on every fold). 

The `pool` can be any instance of `CategoricalArray`, `CategoricalValue` or `CategoricalPool`. [Elsewhere](https://github.com/JuliaAI/MLJXGBoostInterface.jl/blob/4456438d134c20207c0c8383afb14d9f7bfa816c/src/MLJXGBoostInterface.jl#L154), we have just grabbed the first element of the target `y` (a `CategoricalValue`) in fit - let's call it `first_target_element`, and passed that along with the other learned parameters (you call this `model`, the docs  call this the `fitresult`) to make it available to `predict` where the `UnivariateFinite` constructor appears. I suggest you do that. 

Now that I think about it, it may be that your `reformat` operation is destroying this important information (when transforming `y`). So you may need to address the issue there first, so that the output of `reformat` includes the `first_target_element`. Something like

`reformat(model, X, y) -> (X_py, y_py, first_target_element)`

Adjust the signatures of `fit` and `selectrows` accordingly.

ablaom commented:
Perhaps this is helpful:

```julia
julia> v = ["X", "Y", "Z", "Z"] |> categorical
4-element CategoricalArrays.CategoricalArray{String,1,UInt32}:
 "X"
 "Y"
 "Z"
 "Z"

julia> probs = UnivariateFinite(["Z", "X"], [0.1 0.9; 0.2 0.8], pool=v[1])
2-element UnivariateFiniteVector{Multiclass{3}, String, UInt32, Float64}:
 UnivariateFinite{Multiclass{3}}(X=>0.9, Z=>0.1)
 UnivariateFinite{Multiclass{3}}(X=>0.8, Z=>0.2)

julia> pdf(probs, levels(v))
2×3 Matrix{Float64}:
 0.9  0.0  0.1
 0.8  0.0  0.2
```

ablaom commented:
Oh, and can I request that that `fit` not be so verbose. Preferably `verbose <= 1` is silent but at present `verbosity` seems to have no effect at all.

ablaom commented on src/mlj_catboostclassifier.jl:
```diff
@@ -81,20 +81,24 @@ function MMI.fit(mlj_model::CatBoostClassifier, verbosity::Int, data_pool)
-    preds = MMI.UnivariateFinite(classes, pyconvert(Array, py_preds); pool=missing)
+    preds = MMI.UnivariateFinite(classes, pyconvert(Array, py_preds))
     return preds
 end
```
The other option is to do `pool=y_first`. I'm  wondering about whether `classes` as defined here will necessarily have the same order as `classes` defined previously. The important thing is for the ordering in the first two arguments on the `UnivariateFinite` constructor be consistent. Maybe the original definition of `classes` is safer here??

ablaom replied:
Actually, I think you *must* use my suggestive alternative here because the new `classes` will not always have the same length as the old python `classes`. I've just found new fails in the integration tests that suggested this to me.

ablaom replied:
So, to be clear, use `classes` as defined  before your recent commits, but change `pool=missing` to `pool=y_first`.

tylerjthomas9 replied:
Just updated it to use the old method for `classes` with `pool=y_first`

ablaom replied:
Resolved.

ablaom commented:
Good news. If you just add these two missing methods, then all the level-3 MLJIntegrationTests pass, for both the classifier and the regressor:

```julia
function MMI.selectrows(::CatBoostClassifier, I, data_pool)
    py_I = numpy.array(numpy.array(I .- 1))
    return (data_pool.slice(py_I), )
end

function MMI.selectrows(::CatBoostClassifier, I::Colon, data_pool)
    return (data_pool, )
end
```

These are needed so we can subsample data for use in `predict` methods, which will not involve the extra `y_first` piece that `fit` gets.

tylerjthomas9 commented:
> Good news. If you just add these two missing methods, then all the level-3 MLJIntegrationTests pass, for both the classifier and the regressor:

Done! I just expanded the types for the regressor one back to both models

ablaom commented on src/MLJCatBoostInterface.jl:
```diff
@@ -0,0 +1,192 @@
+#####
+##### MLJ docstring, metadata
+#####
+include("mlj_docstrings.jl")
```
Move this `include` so that it appears *after* the package metadata declarations. This will fix the "unknown" that's getting interpolated where "CatBoost.jl" should appear:

```julia
help?> CatBoostClassifier
search: CatBoostClassifier

  CatBoostClassifier

  A model type for constructing a cat boost classifier, based on unknown.jl (unknown), and
  implementing the MLJ model interface.

  From MLJ, the type can be imported using
...
```

tylerjthomas9 replied:
This should be fixed. I now get this: 

```
help?> CatBoostClassifier
search: CatBoostClassifier

  CatBoostClassifier

  A model type for constructing a CatBoost classifier, based on CatBoost.jl.jl (https://github.com/beacon-biosignals/CatBoost.jl), and implementing the MLJ model
  interface.

  From MLJ, the type can be imported using

  CatBoostClassifier = @load CatBoostClassifier pkg=CatBoost.jl
```

ablaom replied:
resolved

ablaom commented on src/mlj_docstrings.jl:
```diff
@@ -0,0 +1,151 @@
+
+
+# Report
+
```
```suggestion
The fields of `report(mach)` are:
```

tylerjthomas9 replied:
Done!

ablaom commented on src/mlj_docstrings.jl:
```diff
@@ -0,0 +1,151 @@
+
+- `predict(mach, Xnew)`: probabilistic predictions of the target given new
+  features `Xnew` having the same scitype as `X` above.
+
```
Please add something like this:

```suggestion

# Accessor functions

- `feature_importances(mach)`: return vector of feature importances, in the form of  `feature::Symbol => importance::Real` pairs

```

ablaom reviewed (approved):
Just a few comments on the docstring for the classifier. The same would apply to the regressor. 

Otherwise good to go.

@tylerjthomas9 Thanks for this valuable contribution and your perseverance.

ablaom commented on src/MLJCatBoostInterface.jl:
```diff
@@ -0,0 +1,193 @@
+
+    return new_model, cache, report
+end
+
```
Sorry, one last thing would be nice to include to save users having to provide it in MLJ's `IteratedModel` wrapper:

```suggestion
MMI.iteration_parameter(::Type{<:CatBoostModels}) = :iterations
```

tylerjthomas9 replied:
Done!

ablaom commented:
Great. Here's a short demo to show what all this hard work helps you do:

```julia
using CatBoost.MLJCatBoostInterface
using MLJBase
using MLJIteration

X, y = @load_iris
model = CatBoostClassifier()
iterated_model = IteratedModel(model, retrain=true)

iterated_model.controls # show default controls
# 5-element Vector{Any}:
#  Step(10)
#  Patience(5)
#  GL(2.0)
#  TimeLimit(Dates.Millisecond(108000))
#  InvalidValue()

mach = machine(iterated_model, X, y) |> fit!
# [ Info: Training machine(ProbabilisticIteratedModel(model = CatBoostClassifier(iterations = 1000, …), …), …).
# [ Info: No iteration parameter specified. Using `iteration_parameter=:(iterations)`. 
# [ Info: No measure specified. Using `measure=LogLoss(tol = 2.220446049250313e-16)`. 
# [ Info: final loss: 0.6239726325698093
# [ Info: Stop triggered by Patience(5) stopping criterion. 
# [ Info: Retraining on all provided data. To suppress, specify `retrain=false`. 
# [ Info: Total of 480 iterations. 
```

For more, see [MNIST Flux example](https://github.com/FluxML/MLJFlux.jl/blob/dev/examples/mnist/notebook.ipynb) and the [official docs](https://alan-turing-institute.github.io/MLJ.jl/dev/controlling_iterative_models/).

Ping me when a new release is tagged and I will updated the MLJ model registry.

tylerjthomas9 commented:
@ericphanson How does this PR look to you? The core functionality should remain the same, and nothing that currently works should break. However, there is now a `CatBoost.MLJCatBoostInterface` module that implements the MLJ interface. The codecov has dropped, but from what I can see, it is just from the MLJ interface.

ericphanson commented on Project.toml:
```diff
@@ -4,22 +4,25 @@ authors = ["Beacon Biosignals, Inc."]
 version = "0.2.0"
```
```suggestion
version = "0.3.0"
```

ericphanson reviewed (approved):
This looks good to me. I think it is technically breaking since we had documented certain functions return DataFrames and they no longer do. However I think it's totally fine to make a breaking release here. So I think we can just bump the release & merge/tag.
